spi-proposal.md

```diff
@@ -0,0 +1,399 @@
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_BEGIN_TRANSACTION (0x01)
+3:  deviceId | channel    (bits 2-6: deviceId, bits 0-1: channel)
+4:  bitOrder | dataMode   (bit 0: bitOrder, bits 1-2: dataMode (0-3))
+5:  clockSpeed            (bits 0 - 6)
+6:  clockSpeed            (bits 7 - 14)
+7:  clockSpeed            (bits 15 - 21)
+8:  clockSpeed            (bits 22 - 28)
+9:  clockSpeed            (bits 29 - 32)
+10  csPin                 [optional] (0-127) The chip select pin number
+11: END_SYSEX
+```
+
+### SPI_END_TRANSACTION
+
+Signals the end of a transaction. Must be sent before starting communication
+with another SPI device.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_END_TRANSACTION   (0x02)
+3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### SPI_TRANSFER
+
+Write, read or read and write one or more bytes.
+
+The user can continue sending as many transfer messages as necessary, even
+alternating between read-only, write-only and read-write as necessary within
+a single SPI transaction.
+
+When setting `0` (read/write) as the `transferOption`, a byte must be written for
+every byte received. Write `0x00` when you only need a byte returned. The user
+must then extract data from the array of bytes returned in the `SPI_REPLY`
+message.
+
+If a CS pin is specified in the `SPI_BEGIN_TRANSACTION` message, it will be used
+according to the behavior specified by `pinOptions` (byte 4). This only covers
+common cases, with the default being set CS to `LOW `before sending data and set
+to `HIGH` after the last byte has been sent. Using the `pinOptions` bits, it
+also enables the user to set the CS pin value only on start (bit 4) or only on
+end (bit 5) or ignore entirely (bit 2) in order to send data that requires
+multiple `SPI_TRANSFER` messages. Another option (bit 6) allows the user to
+specify whether or the CS pin should be toggled at the end of the message
+(typical SPI behavior) or between each byte sent (edge case).
+
+*Note: It may be useful to separate pinOptions and transferOptions to handling
+additional use cases as they arise.*
+
+*Note: TBD if a single `SPI_TRANSFER` message is sufficient or if we need
+additional transfer messages such as `SPI_TRANSFER_16`.*
+
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_TRANSFER          (0x03)
+3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  pinOptions | transferOptions
+                          bits 0-1: transfer options
+                            0 = read/write,
+                            1 = read-only,
+                            2 = write-only)
+                            if read-only, a value of 0x00 will be written.
+                            if read/write or read-only, respond with `SPI_REPLY`
+                          bits 2-4 : pin options
+                            bit 2: 0 = no CS, 1 use CS
```

Comment:
I need to flip the logic for bit 2. Since having a CS pin is likely the most common action, this should read:
```
0 = CS Active
1 = CS Inactive
```

The default configuration across bits 2 - 6 should be `0`.